registry/quilt_server/views.py

```diff
@@ -382,6 +354,97 @@ def wrapper(*args, **kwargs):
         return wrapper
     return innerdec
 
+@app.route('/login', methods=['POST'])
+@api(require_anonymous=True, require_login=False, schema=USERNAME_PASSWORD_SCHEMA)
+@as_json
+def login_post():
+    try:
+        data = request.get_json()
+        username = data.get('username')
+        password = data.get('password')
+        if try_login(username, password):
+            token = issue_token(username)
+        else:
+            return {'error': 'Login attempt failed'}
+
+        db.session.commit()
+        return {'token': token}
+    except Exception:
+        return {'error': 'Login failed.'}, requests.codes.unauthorized
+
+CORS(app, resources={"/login": {"origins": "*", "max_age": timedelta(days=1)}})
+
+@app.route('/activate/<link>')
+def activate_endpoint(link):
+    return activate_response(link)
+
+CORS(app, resources={"/activate/*": {"origins": "*", "max_age": timedelta(days=1)}})
+
+@app.route('/reset_password', methods=['POST'])
+@api(require_anonymous=True, require_login=False, schema=PASSWORD_RESET_SCHEMA)
+@as_json
+def reset_password_endpoint():
+    return reset_password_response()
+
+CORS(app, resources={"/reset_password": {"origins": "*", "max_age": timedelta(days=1)}})
+
+@app.route('/api-root')
+@api()
+@as_json
+def apiroot():
+    return {'is_staff': g.auth.is_admin, 'is_active': g.auth.is_active,
+            'email': g.auth.email, 'current_user': g.auth.user}
+
+CORS(app, resources={"/api-root": {"origins": "*", "max_age": timedelta(days=1)}})
+
+@app.route('/register', methods=['POST'])
+@as_json
+def register_endpoint():
+    data = request.get_json()
+    if app.config['DISABLE_SIGNUP']:
+        raise ApiException(400, "Signup is disabled.")
+    username = data['username']
+    password = data['password']
+    email = data['email']
+    _create_user(username, password=password, email=email)
+    db.session.commit()
+    return {}
+
+CORS(app, resources={"/register": {"origins": "*", "max_age": timedelta(days=1)}})
+
+@app.route('/api/refresh', methods=['POST'])
+@api()
+@as_json
+def refresh():
+    token_str = request.headers.get(AUTHORIZATION_HEADER)
+    revoke_token_string(token_str)
+    token = issue_token(g.auth.user)
+    db.session.commit()
+    return {'token': token}
+
+@app.route('/logout', methods=['POST'])
+@api()
+@as_json
+def logout():
+    data = request.get_json()
+    token = data['token']
```

Comment:
Compiler depends on refresh, so it needs to be backward compatible. Logout is only depended on by catalog, so it can conform to convention, which is JSON.

Reply:
I was mistaken, compiler uses `/api/token`. Being an `@api` endpoint means that the token is already present and valid, so I will change logout to use the header for consistency.